v2/internal/reconcilers/arm/arm_client_factory.go

```diff
@@ -6,8 +6,11 @@
 package arm
 
 import (
+	"context"
+
 	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
 	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"k8s.io/client-go/tools/record"
 )
 
-type ARMClientFactory func(genruntime.ARMMetaObject) *genericarmclient.GenericClient
+type ARMClientFactory func(context.Context, genruntime.ARMMetaObject, record.EventRecorder) (*genericarmclient.GenericClient, error)
```

Comment:
I actually think it's OK for you to turn this into a struct like you had before (at least, I think you had that before).

Then you could pass the `record.EventRecorder` to that struct and do the event emission in the bit of code that constructs a new client (in `GetClient`).

Thoughts?

Reply:
Basically I am not sure we need both `armClients` _and_ `armClientFactory`. They seem two things that are tackling the same or very similar problem?

Am I missing the need for both?

Reply:
`armClientFactory` is just a factory method which gets passed down to the reconciler so that each time we reconcile a resource, we can have values put into the factory method, which then has access to the to the `armClientCache`.

Reply:
I used eventRecorder as a parameter so that we don't have to call `mgr.getRecorder` and have another instance of recorder, rather we can just pass it [here](https://github.com/Azure/azure-service-operator/pull/2559/files#diff-48b587518825e853a2effad5b6a94c94327f2a645f701d9f71c18d954c0643ecR102) and use it.

Reply:
As we're using `armClientFactory` in `kubeTestContext` as well with a completely different logic, I think for now, as we discussed, its better to keep both and tackle this in a separate PR